daemon/api/dependencies.py

```diff
@@ -77,27 +85,95 @@ def localpath(self) -> Path:
                 detail=f'File `{self.filename}` not found in workspace `{self.workspace_id}`',
             )
 
-    @cached_property
-    def port_expose(self) -> str:
+    @property
+    def newfile(self) -> str:
+        """return newfile path in format
+        `<root-workspace>/<workspace-id>/<flow-id>_<original-filename>`
+
+        :return: return filepath to save flow config in
         """
-        Sets `port_expose` for the Flow started in `partial-daemon`.
-        This port needs to be exposed before starting `partial-daemon`, hence set here.
-        If env vars are passed, set them in current env, to make sure all values are replaced.
-        Before loading the flow yaml, change CWD to workspace dir.
+        return get_workspace_path(self.workspace_id, f'{self.id}_{self.filename}')
 
-        :return: port_expose
+    def load_and_dump(self) -> None:
+        """
+        every Flow created inside JinaD lives inside a container. It is important to know the
+        list of ports to be published with localhost before actually starting the container.
+
+        1. `load` the flow yaml here.
+            - yaml is stored in `workspace` directory, so we'll `cd` there
+            - yaml might include env vars. so we'll set them (passed via query params)
+        2. `build` the Flow so that `gateway` gets added.
+            - get the list of ports to be published (port_expose, port_in, port_out, port_ctrl)
```

Comment:
Just FYI: For GprcDataRuntime we dont need port_out and port_ctrl - But it also does not hurt to expose them as well for now. Eventually we might want to be a bit tighter here though

Reply:
right. We'll be strict here with port mapping, as we improve on it.